Map code coverage from transformers

jwbay opened the pull request:
**Summary**
With this PR transformers can supply source maps to Jest and see code coverage mapped against the original sources in all reports without any further processing.

Example transformer - https://github.com/jwbay/jestcoveragetest/blob/master/preprocessor.js#L30-L33

The support for threading source maps through like this was only recently added to the istanbul-lib-x packages.

**Test plan**
`jest`

There is a test repo here that I've been using to verify behavior. Screenshots are from that. https://github.com/jwbay/jestcoveragetest

### Before:
![image](https://cloud.githubusercontent.com/assets/1320447/21552467/b751cccc-cdcf-11e6-80b6-f28ea9382e44.png)

### After:
![image](https://cloud.githubusercontent.com/assets/1320447/21552499/f317e93a-cdcf-11e6-9f58-2f91af2f88e3.png)

It's not completely perfect but I think it's the closest we can get for now -- I think the trailing braces are an issue with fidelity of the supplied source maps. It seems inline with the results of remap-istanbul, which is basically the gold standard.

TODO:
- [x] Get https://github.com/istanbuljs/babel-plugin-istanbul/pull/75 landed and released
- [x] Accept non-inline source maps from transformers
- [x] Add integration test to assert on remapped coverage
- [x] Explore external sources
- [x] Explore on-the-fly remapping
- [x] Add option to explicitly disable remapping

cc @kulshekhar, @kwonoj, and @Igmat who have been working with coverage remapping in ts-jest. If you all have any test cases it would be good to see if this handles them.

kwonoj commented:
One of the suffering point of external code coverage mapping is jest does own caching mechanism and does not invoke preprocessor if a file has not changed. Is this change resolves those issue? Just looking at glimpse it relies on existing preprocessor pipeline.

jwbay commented:
Short answer, yes it resolves that. This approach does rely on the source map having the original sources inlined into the sourcesText property. The source map supplied to the instrumenter is attached to the coverage object by istanbul and written out alongside the covered code, which is what Jest caches.

dmitriiabramov commented on packages/jest-runtime/src/transform.js:
```diff
@@ -293,6 +294,30 @@ const transformSource = (
+
+  if (shouldTransform(filename, config)) {
+    if (transform.canInstrument && typeof transform.getEmptyCoverage === 'function') {
+      return transform.getEmptyCoverage(content, filename, config, {
```
does that delegate getting the empty coverage object to a transformer if it implements it?

jwbay replied:
Yes, exactly. Not sure how important visual accuracy is for uncovered code, but it can look pretty busted without this, so you'd probably get a couple issues now and then. Also prevents the branch/etc count from jumping up/down depending on whether the file executed or not.

dmitriiabramov commented on scripts/build.js:
```diff
@@ -69,7 +69,7 @@ function buildFile(file, silent) {
   const destPath = path.resolve(packageBuildPath, relativeToSrcPath);
 
-  spawnSync('mkdir', ['-p', path.dirname(destPath)]);
+  mkdir.sync(path.dirname(destPath));
```
did it fail on win or something? :)

jwbay replied:
Yes 😅 Since mkdirp was already in jest's dependencies I figured it wouldn't hurt.

dmitriiabramov commented on packages/jest-cli/src/reporters/CoverageReporter.js:
```diff
@@ -59,6 +62,7 @@ class CoverageReporter extends BaseReporter {
     runnerContext: RunnerContext,
   ) {
     this._addUntestedFiles(config, runnerContext);
+    const { map, sourceFinder } = this._sourceMapStore.transformCoverage(this._coverageMap);
```
which part is actually reading the source maps? and what happens if source maps are not available?

cpojer replied:
nit: code style, this should be `{map, sourceFinder}`.

jwbay replied:
@dmitriiabramov istanbul-lib-source-maps has code that plucks the source map from the coverage object as part of this transformCoverage call. If no source map is tacked onto the object, it's not mapped.

@cpojer Will fix shortly!

cpojer commented on packages/jest-cli/src/reporters/CoverageReporter.js:
```diff
@@ -75,7 +79,7 @@ class CoverageReporter extends BaseReporter {
 
       reporter.addAll(coverageReporters);
-      reporter.write(this._coverageMap);
+      reporter.write(map, { sourceFinder });
```
code style too.

cpojer commented on packages/jest-runtime/src/transform.js:
```diff
@@ -342,3 +367,4 @@ module.exports = (
 
 module.exports.EVAL_RESULT_VARIABLE = EVAL_RESULT_VARIABLE;
 module.exports.transformSource = transformSource;
+module.exports.generateEmptyCoverage = generateEmptyCoverage;
```
please add an empty line at the end of this file.

cpojer commented:
oh wow, this is indeed cool! That will make ts-jest a lot nicer I'm sure.

jwbay commented:
@dmitriiabramov I can't seem to reply directly to your question with the images, but yes, it's expected. The branch/statement and sometimes function count can be different in the original source vs. the transformed/covered code if a compiler injects new branches/statements/etc. When this is reconciled during the remapping, the percentages are changed as a result. Whether that's correct behavior seems to be an open question with lots of opinions :)

dmitriiabramov commented:
and what about the performance? is it much slower to do the mapping?

jwbay commented:
I'll do some performance testing. Hopefully it's not a huge hit, but I'm sure it's not free. It's going to be doing this extra work per file with coverage - https://github.com/istanbuljs/istanbul-lib-source-maps/blob/master/lib/transformer.js#L84.

Also, I found a way to make this work with babel that should make the transformers do far less work. They won't need to take over instrumentation to see remapped coverage and we won't need to add getEmptyCoverage to the transformer API. It needs [this PR](/istanbuljs/babel-plugin-istanbul/pull/75) merged and a different kind of transformer API change. Inline source maps will just work and we could limit this to transformers that support them, but I'd like to make `transformer.process` return either a content string as it does today, or an object like `{ code: string, sourceMap: Object }`.

cpojer commented:
This will only affect performance when doing `jest --no-cache --coverage`, right? Can you take a few samples before and after and give us some numbers? This may make sense on the Jest repo itself.

jwbay commented:
@cpojer It will affect performance on every run that does --coverage if source maps are present. Cache doesn't come into play unless we're talking about cost of generating source maps, which depends on the transformer.

Here are some perf numbers. I opted for synthetic tests to have more control and test project sizes. Each file has 60 statements, 50 branches, 20 functions, and 50 lines. Each row is an average of 5 runs. 'Coverage time' is time spent in the onRunComplete method of CoverageReporter.

Without mapped coverage (master branch)

| Test file count | Total time | Coverage time
| ---: | ---:| ---:
| 10 | 3.00s | 0.23s
| 100 | 7.95s | 0.47s
| 1000 | 55.42s | 2.90s

With mapped coverage and source maps

| Test file count | Total time | Coverage time
| ---: | ---:| ---:
| 10 | 3.08s | 0.28s
| 100 | 8.67s | 0.76s
| 1000 | 57.18s | 4.43s

With mapped coverage, but no source maps supplied

| Test file count | Total time | Coverage time
| ---: | ---:| ---:
| 1000 | 55.61s | 2.86s

jwbay commented:
Another concern is memory. If the source map has its source content inlined, you could end up with the source code text of your entire project held in memory by the main Jest process. This could be bad 😅 I'll look at whether it's possible to remap on the fly and clean up after.

dmitriiabramov commented:
@jwbay i don't think it's possible to clean up on the fly, because you don't know if there's going to be more coverage coming for a specific file until the end of the test suite. But we definitely need to make sure we only collect source map for files we intend to cover and that we provide an option to disable mapping (at fb souce maps will be gigabytes of data 😄 )

jwbay commented:
I meant more cleaning up the source map than the coverage. When a single test result comes in with a source map attached, I think we might able to remap there and delete the source map, so we're only ever merging mapped coverage and the source map doesn't live past its usefulness. This would hurt runtime if a file is executed multiple times across tests, though. You could be mapping against the same source map multiple times instead of once at the end of the run.

Completely turning off source maps is going to be difficult with [this PR](/istanbuljs/babel-plugin-istanbul/pull/75) in its current form. If source maps are inlined with the content, they'll get picked up by Babel and attached to coverage by the plugin. I'll add some kind of flag to prevent that behavior since it's pretty memory intensive to not have an off switch.

dmitriiabramov commented:
i'm super excited about this RP, i'll try to test in on the internal FB repo and see how much memory it eats

jwbay commented:
This is ready for review. The babel plugin update is released as an RC at the moment. Once that's promoted all the boxes will be ticked.

Strate commented:
Could this PR helps to debug original code with source maps?

jwbay commented:
@Strate No, this PR only deals with code coverage. Debugging original source code from a transformer is already possible. You'll need to check whether your IDE supports debugging with source maps.

Strate commented:
> Debugging original source code from a transformer is already possible.

Didn't find any way to return source maps from transformer in docs, is it undocumented feature?

jwbay commented:
Inline source maps don't need to be returned separately.

bcoe commented:
@jwbay your work should officially be out the door now in:

`+ babel-plugin-istanbul@3.1.1` was blocked by an issue with `istanbul-lib-instrument`.

dmitriiabramov commented:
this is currently blocked by https://github.com/babel/babel/issues/5081
@bcoe maybe you can give some input on this one too!

bcoe commented:
@dmitriiabramov happy to give feedback 👍 and am happy to help move forward any issues on the istanbul repos that help.

jwbay commented:
Rebased to resolve conflicts. @orta Danger was insisting I add @flow to a file that's not compiled by babel, so I worked around it [here](https://github.com/facebook/jest/pull/2290/commits/cf49749c81980f846022e0ff3cc0c99e9255739bhere). Could this be handled better?

jwbay commented:
Rebased more conflicts away. Also squashed some of the more ancient commits because the list was getting pretty long.

@dmitriiabramov @cpojer Is there any chance this could get landed without testing on the FB codebase if we flip the new mapCoverage config option from on by default to off? I've hidden all the new code I can behind that flag to avoid performance regressions if it's off.

The babel PR to fix FB's upgrade situation is not moving too quickly, but even once it's landed it still needs to be released.

dmitriiabramov commented:
hey @jwbay 
i was actually able to rollback some dependencies and make jest run on fb without freezing the process. I'll try to test your PR agains fb codebase on thursday and let you know if everything works fine!

jwbay commented:
@dmitriiabramov You can hold off on the FB test if you'd like -- I found a large-ish project to test this on and it choked hard on tests with large dependency graphs that `require` lots of files. I was hoping to gate and drop stuff at the TestResult boundary if needed, but it's not even getting there. The worker processes run out of memory. Source maps are big. 😅  I'll need to look at storing them on disk in the transform cache.

dmitriiabramov commented:
@jwbay i tried it on fb codebase today, but couldn't get it to work with our custom preprocessor. The source maps were being generated, but i couldn't map them correctly. I'll try to debug it.

I think it's still cool to have this option available though, but probably disable it by default. Smaller projects will still benefit from this feature

dmitriiabramov commented:
@jwbay do you think passing `compact` option to babel when mapping coverage will help?
it shouldn't matter any more since we're mapping line numbers anyway

jwbay commented:
One thing that might be happening with your preprocessor is that if it takes over instrumentation, the injection of the source map into the coverage map doesn't necessarily happen. To do source map injection, babel-plugin-istanbul either needs to be passed a source map explicitly, or the input file needs to have an inline source map present. This basically requires a two-pass transform.

For an instrumenting transformer like babel-jest where instrumentation happens in the same pass as compilation, I don't think a remapping step should be needed. I think coverage fidelity there is more an issue of trying to get babel-plugin-istanbul running first before any other babel plugins move code around.

As far as getting this in for smaller projects, that sounds good to me!

jwbay commented:
@dmitriiabramov You might get a kick out of this -- turns out the test lockups I assumed were caused by memory pressure were actually caused by [the same issue](https://github.com/babel/babel/issues/5081) you found. In my case it was a vendored library that ended up being a 3MB string after instrumentation. It was a coincidence that a couple of the largest dependency graphs in the app happened to hit that file.

I switched the implementation to store source maps on disk anyway, though. It's much safer performance wise. The cost is now concentrated as CPU time at the end of the run instead of spread out as memory during the run (and costing CPU time at the end anyway). A project with 3K files takes about 1500ms to remap, and a project with a couple hundred files takes 300ms.

dmitriiabramov commented on packages/jest-runtime/src/transform.js:
```diff
@@ -276,21 +280,55 @@ const transformSource = (
+    const sourceMapFilePath = cacheFilePath + '.map';
+    writeCacheFile(sourceMapFilePath, sourceMapContent);
+    result +=
+      `\n;global.__coverage__['${escapePathForJavaScript(filename)}']` + 
```
isn't `__coverage__` name configurable from istanbul? we can potentially break coverage collection.
i also think that this wrapping logic deserves to be a separate function and `inputSourceMapPath` property name should be extracted into a separate constant, since it's something we introduce to istanbul coverage object

bcoe replied:
@dmitriiabramov well in practice, it can pretty much be assumed that `__covearge__` is the variable (I don't know that I've seen many deviations), this variable can be configured:

https://github.com/istanbuljs/istanbul-lib-instrument/blob/b08e4f55d60a43d542add9b09b3bf52a1d88ce6f/src/instrumenter.js#L13

jwbay replied:
Yeah, the more I look at this the less I like it. To do this correctly, we could try and update the istanbul-lib-x packages to either overload the inputSourceMap property to accept a path to a source map as well as a source map proper, or add a new inputSourceMapPath property. Either one would be fairly straightforward, it would just need to touch two or three packages. @bcoe what do you think?

dmitriiabramov commented on types/TestResult.js:
```diff
@@ -11,12 +11,23 @@
+  statementMap: { [statementId: number]: any },
+  branchMap: { [branchId: number]: any },
+  inputSourceMap?: Object,
+  inputSourceMapPath?: string,
```
is this istanbul shape, or are we adding sourcemap properties to it?

jwbay replied:
`inputSourceMap` is official. `inputSourceMapPath` is not (yet).

dmitriiabramov commented on types/Transform.js:
```diff
@@ -11,6 +11,11 @@
 
+export type TransformedSource = {|
+  content: string,
+  sourceMap: ?Object | string,
```
is that the way typescript compiler returns it by default?
if not, should we rename it to babel's defaults?
i think it was `{code, map}` shape

jwbay replied:
No, typescript returns something different. I was trying not to play favorites with compilers *too* much. I'll change to babel's shape.

cpojer commented:
I'd prefer to hold off until after Jest 19 with this one.

jwbay commented:
@dmitriiabramov I haven't really tried to remap babel output for the reason you mentioned. I'll see if I can add a test.

jwbay commented:
@dmitriiabramov ecbde2083e384c1956f300a3d2dbebbb161f7fc3 switches the approach from source map info hitching a ride on the global coverage object to being more of a first class citizen in jest-runtime. It required some function signature changes, but disconnecting source maps from coverage a bit has some benefits. It sets Jest up for remapping stacktraces for errors, and jest-editor-support might be able to do some cool stuff with source maps exposed. Also, we can now map coverage for untested files, which doesn't seem too important but will probably save Jest a few github issues now and then.

jwbay commented:
As far as compiler support, it doesn’t look good for babel. I tested four compilers: babel, typescript, buble, and coffeescript. The [testbed](/jwbay/jestcoveragetest) is updated with the transformers I used.

* It seems Babel is doing something with its source maps that the `source-map` package doesn’t like. It ends up returning a lot of wrong/invalid positions. Sometimes it gets things right, but not often. I tried `retainLines` to help it along but it didn’t do much.
* Typescript is pretty solid. It’s not perfect, but I’ve not seen things go horribly wrong.
* Buble is also solid, which is impressive since it apparently does string transformations instead of an AST transform.
* Coffeescript is basically all over the place.

I debugged a bit and it looks like istanbul-lib-source-maps isn’t doing anything weird or wrong. It seems to come down to the [source-map](https://github.com/mozilla/source-map) package returning odd source positions for source maps from some compilers. Given all this, I’ve added a disclaimer of sorts to the docs for the flag.

bookman25 commented:
Just ran into a good use case for this.  By having jest provide the source map information, it would allow jest-editor-support to leverage that info for generating code coverage overlays.

dmitriiabramov commented:
i tested it locally and it works really nice!
no luck with www, but it's because of our monstrous transform pipeline (i don't think it's a blocker since it'll not affect www and we won't be able to use it right now anyway)

dmitriiabramov commented:
@jwbay thank you so much for working on it! and sorry it took so long :)
This is going to be a game changer for TS projects!
